### PR TITLE
Method to get fully-qualified names of types; run mypy on tests

### PR DESCRIPTION
**Pull request by rgabbard**

### CODE REVIEW

**Comment by berquist** [comment]
CI hasn't finished yet, but `make test` is giving me
```
nosetests --exe
........F......................./Users/berquist/opt/apps/python/anaconda/envs/vistautils/lib/python3.7/zipfile.py:1470: UserWarning: Duplicate name: '__keys'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
...............................................................
======================================================================
FAIL: test_slots_with_no_parent (tests.test_generic_bug.TestGenericBug)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/berquist/development/vistautils/tests/test_generic_bug.py", line 30, in test_slots_with_no_parent
    class Foo4(Generic[T]):
AssertionError: TypeError not raised

----------------------------------------------------------------------
Ran 95 tests in 1.805s

FAILED (failures=1)
make: *** [test] Error 1
```

**Comment by rgabbard** [comment]
@berquist : Strange - that code was untouched in this commit. Can you see if the same test failure happens on `master` for you?

**Comment by berquist** [comment]
Yes, it fails on master as well. Not sure if it matters, but here are the contents of my (fresh) env:
```
# packages in environment at /Users/berquist/opt/apps/python/anaconda/envs/vistautils:
#
# Name                    Version                   Build  Channel
alabaster                 0.7.12                    <pip>
appdirs                   1.4.3                     <pip>
astroid                   2.1.0                     <pip>
attrs                     18.2.0                    <pip>
Babel                     2.6.0                     <pip>
black                     18.9b0                    <pip>
ca-certificates           2018.10.15           ha4d7672_0    conda-forge
certifi                   2018.10.15            py37_1000    conda-forge
chardet                   3.0.4                     <pip>
Click                     7.0                       <pip>
coverage                  4.5.2                     <pip>
docutils                  0.14                      <pip>
flake8                    3.5.0                     <pip>
frozendict                1.2                       <pip>
idna                      2.7                       <pip>
imagesize                 1.1.0                     <pip>
immutablecollections      0.1.2                     <pip>
incremental               17.5.0                    <pip>
isort                     4.3.4                     <pip>
Jinja2                    2.10                      <pip>
lazy-object-proxy         1.3.1                     <pip>
libffi                    3.2.1                hfc679d8_5    conda-forge
MarkupSafe                1.1.0                     <pip>
mccabe                    0.6.1                     <pip>
mypy                      0.620                     <pip>
ncurses                   6.1                  hfc679d8_1    conda-forge
nose                      1.3.7                     <pip>
nose-exclude              0.5.0                     <pip>
openssl                   1.0.2p               h470a237_1    conda-forge
packaging                 18.0                      <pip>
pip                       18.1                  py37_1000    conda-forge
pycodestyle               2.3.1                     <pip>
pyflakes                  1.6.0                     <pip>
Pygments                  2.3.0                     <pip>
pylint                    2.1.1                     <pip>
pyparsing                 2.3.0                     <pip>
python                    3.7.1                h46c1a51_0    conda-forge
pytz                      2018.7                    <pip>
PyYAML                    3.12                      <pip>
readline                  7.0                  haf1bffa_1    conda-forge
requests                  2.20.1                    <pip>
setuptools                40.6.2                   py37_0    conda-forge
six                       1.11.0                    <pip>
snowballstemmer           1.2.1                     <pip>
sortedcontainers          2.1.0                     <pip>
Sphinx                    1.8.2                     <pip>
sphinx-autodoc-typehints  1.5.1                     <pip>
sphinxcontrib-websupport  1.1.0                     <pip>
sqlite                    3.25.3               hb1c47c0_0    conda-forge
tk                        8.6.9                ha92aebf_0    conda-forge
toml                      0.10.0                    <pip>
towncrier                 18.6.0                    <pip>
typed-ast                 1.1.0                     <pip>
typing-extensions         3.6.6                     <pip>
urllib3                   1.24.1                    <pip>
wheel                     0.32.3                   py37_0    conda-forge
wrapt                     1.10.11                   <pip>
xz                        5.2.4                h470a237_1    conda-forge
zlib                      1.2.11               h470a237_3    conda-forge
```

**Comment by berquist** [comment]
Anyway, I'll ignore that for right now and look at the PR contents.

**Comment by rgabbard** [comment]
@berquist : Aha, that test was looking for a failure which depends on a bug present in Python 3.6 which is fixed in 3.7.  I'll mark it as `@skip`, since it shouldn't be running anyway.

File: tests/test_generic_bug.py [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
-from unittest import TestCase
+from unittest import TestCase, skip
 
 from attr import attrs
 
 T = TypeVar("T")
 
 
 # this test is to track https://github.com/python-attrs/attrs/issues/313
+# we skip this because we don't want to crash on Python 3.8, where this bug appears to be gone
```
Comment (berquist) [inline tests/test_generic_bug.py]: This should be 3.7, not 3.8.

Comment (rgabbard) [reply tests/test_generic_bug.py]: Fixed

**Pull request merged**